flang/lib/Lower/SymbolMap.cpp

```diff
@@ -0,0 +1,170 @@
+  if (box.getLBounds().size()) {
+    os << ", lbounds: [";
+    llvm::interleaveComma(box.getLBounds(), os);
+    os << "]";
+  } else {
+    os << ", lbounds: all-ones";
+  }
+  os << ", shape: [";
+  llvm::interleaveComma(box.getExtents(), os);
+  return os << "]}";
+}
+
+llvm::raw_ostream &fir::operator<<(llvm::raw_ostream &os,
+                                   const fir::CharArrayBoxValue &box) {
+  os << "boxchararray { addr: " << box.getAddr() << ", len : " << box.getLen();
+  if (box.getLBounds().size()) {
+    os << ", lbounds: [";
+    llvm::interleaveComma(box.getLBounds(), os);
+    os << "]";
+  } else {
+    os << " lbounds: all-ones";
+  }
+  os << ", shape: [";
+  llvm::interleaveComma(box.getExtents(), os);
+  return os << "]}";
+}
+
+llvm::raw_ostream &fir::operator<<(llvm::raw_ostream &os,
+                                   const fir::BoxValue &box) {
+  os << "box { addr: " << box.getAddr();
+  if (box.getLen())
+    os << ", size: " << box.getLen();
+  if (box.params.size()) {
+    os << ", type params: [";
+    llvm::interleaveComma(box.params, os);
+    os << "]";
+  }
+  if (box.getLBounds().size()) {
+    os << ", lbounds: [";
+    llvm::interleaveComma(box.getLBounds(), os);
+    os << "]";
+  }
+  if (box.getExtents().size()) {
+    os << ", shape: [";
+    llvm::interleaveComma(box.getExtents(), os);
+    os << "]";
+  }
+  return os << "}";
+}
+
+llvm::raw_ostream &fir::operator<<(llvm::raw_ostream &os,
+                                   const fir::ProcBoxValue &box) {
+  return os << "boxproc: { procedure: " << box.getAddr()
+            << ", context: " << box.hostContext << "}";
+}
+
+llvm::raw_ostream &fir::operator<<(llvm::raw_ostream &os,
+                                   const fir::ExtendedValue &exv) {
+  exv.match([&](const auto &value) { os << value; });
+  return os;
+}
+
+void Fortran::lower::SymMap::addSymbol(Fortran::semantics::SymbolRef sym,
+                                       const fir::ExtendedValue &exv,
+                                       bool force) {
+  exv.match([&](const fir::UnboxedValue &v) { addSymbol(sym, v, force); },
+            [&](const fir::CharBoxValue &v) { makeSym(sym, v, force); },
+            [&](const fir::ArrayBoxValue &v) { makeSym(sym, v, force); },
+            [&](const fir::CharArrayBoxValue &v) { makeSym(sym, v, force); },
+            [&](const fir::BoxValue &v) { makeSym(sym, v, force); },
+            [](auto) {
+              llvm::report_fatal_error(
+                  "box value should not be added to symbol table");
+            });
+}
+
+Fortran::lower::SymbolBox
+Fortran::lower::SymMap::lookupSymbol(Fortran::semantics::SymbolRef sym) {
+  for (auto jmap = symbolMapStack.rbegin(), jend = symbolMapStack.rend();
+       jmap != jend; ++jmap) {
+    auto iter = jmap->find(&*sym);
+    if (iter != jmap->end())
+      return iter->second;
+  }
+  // FIXME BUG: How does this know that the caller is expecting the host
+  // associated symbol?
```

Comment:
I think I want the distinction to be forced on the client though.

`lookupSymbol(x)` logically should lookup the symbol `x` and nothing more.

On the other hand, a `lookupAssociatedSymbol(x)` would lookup either `x` or whatever `x` was associated with. I don't know if there will ever be a case where the former is ever required, but I'm not sure I like conflating these two things by default.

Reply:
I understand your point. All the places I can think of would need to use `lookupAssociatedSymbol` as of today.
Two alternatives are:
a) add a boolean flag to disable the fallback to host associated symbols for the clients where it would not make sense to follow host association (since in my opinion, these should be the weirdo cases).
b) create SymbolBox for HostAssociated symbols in any cases and remove this fallback. That may just be the cleanest solution. Only drawback is that it may lead to a bigger symbol map, but I do not think there should be too many of those symbols where we would create a SymbolBox for nothing (only Blocks/Stmt functions and maybe OpenMp).

Reply:
If the 99% case is that a symbol marked as "host associated" really means the variable it is "host assoicated" with then that semantics should be made explicit in the binding (i.e., (b)). Probing the table with a magic error recovery side effect is muddy. With the current approach, it's unclear if the symbol *must* be in the table, if the lookup is for a particular symbol *only* and where the context should be aware if the symbol shadows another variable, if a symbol isn't found whether the error is that missing symbol or some other missing symbol, etc.